Attaching to multiple networks followup 2

qinqon opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
-->

**What this PR does / why we need it**:
Multiple Network Attachment with bridge CNI (#631)
    
Add blog port post to do follow up on multi network connectivity blog post
but this time with bridge CNI instead of OVS, cluster-network-addons-operator
to install all the kubevirt network operators and kubernetes-nmstate to
configure the bridge at nodes.
    
It also install jekyll-diagrams [1]
    
[1] https://github.com/zhustec/jekyll-diagrams

**Does this PR fix any issue?** _(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)_:

> Fixes #

**Special notes for your reviewer**:
Depends on https://github.com/kubevirt/kubevirt.github.io/pull/637

phoracek commented on _posts/2020-09-23-attaching-to-multiple-networks-followup.markdown:
```diff
@@ -6,7 +6,7 @@ description: This post illustrate configuring secondary interfaces at VMs with a
 navbar_active: Blogs
 pub-date: September 23
```
Bring this up to date, so it appears last in the list of blog posts

maiqueb reviewed (changes requested):
Please take the opportunity to also address:
  - https://github.com/kubevirt/kubevirt.github.io/pull/631#discussion_r503913074
  - https://github.com/kubevirt/kubevirt.github.io/pull/631#discussion_r503918324

RamLavi commented on _posts/2020-09-23-attaching-to-multiple-networks-followup.markdown:
````diff
@@ -0,0 +1,605 @@
+sudo install virtctl /usr/local/bin
+```
+
+Now let's create two VirtualMachineInstances on each node they will have one secondary NIC connected to br0 using the multus configuration for vlan 100, also activate kubemacpool to be sure that mac addresses are unique in the cluster and install the qemu-guest-agent so ip addresses from secondary NICs are reported to VMI so we can inspect them later on.
````
```suggestion
Now let's create two VirtualMachineInstances on each node they will have one secondary NIC connected to br0 using the multus configuration for vlan 100. We will also activate kubemacpool to be sure that mac addresses are unique in the cluster and install the qemu-guest-agent so ip addresses from secondary NICs are reported to VMI and we can inspect them later on.
```

RamLavi commented on _posts/2020-09-23-attaching-to-multiple-networks-followup.markdown:
```diff
@@ -0,0 +1,605 @@
+        mutatevirtualmachines.kubemacpool.io: allocate
+---
+apiVersion: kubevirt.io/v1alpha3
+kind: VirtualMachineInstance
```
if you want KMP to work you need to run vms, not vmis.

RamLavi commented on _posts/2020-09-23-attaching-to-multiple-networks-followup.markdown:
```diff
@@ -0,0 +1,605 @@
+    ipAddresses:
+    - 10.244.196.145/32
+    - fe80::ac2a:27ff:fe59:204/64
+    mac: ae:2a:27:59:02:04
```
notice the macs allocated on `eth0`, `eth1` are not on same range

RamLavi reviewed (changes requested):
we need to change to vms instead of vmis.
KMP allocates mac to vmis via the pod webhook and we currently have a bug there + it's not opted in. plus it's better practice to use vms...

rhrazdil commented on _posts/2020-09-23-attaching-to-multiple-networks-followup.markdown:
```diff
@@ -0,0 +1,605 @@
+> note "Note"
+> There is a known issue related to clusters not having a default gateway making the nncp always to rollback, there are plans to use different approach like checking name servers.
+
+The desired state field follow the nmstate API described at their [awesome doc](https://www.nmstate.io/)
```
```suggestion
The desired state fields follow the nmstate API described at their [awesome doc](https://www.nmstate.io/)
```

rhrazdil commented on _posts/2020-09-23-attaching-to-multiple-networks-followup.markdown:
````diff
@@ -0,0 +1,605 @@
+        type: linux-bridge
+```
+
+We can also check that the `bridge-marker` is working and check resources on nodes:
````
just think that this may sound a bit better
```suggestion
We can also check that the `bridge-marker` is working and check verify on nodes:
```

qinqon commented:
Pushed:
* Addressed Radim and Ram comments
* Fixed set-context issue with deploy.sh from HCO
* Pod network binding set to masquerade since it's the default binding configured at HCO if no interface is specified.

qinqon commented:
Push:
* Fixed VMs interface output now that MAC address comes from kubemacpool.

RamLavi commented on _posts/2020-09-23-attaching-to-multiple-networks-followup.markdown:
```diff
@@ -0,0 +1,618 @@
+> note "Note"
+> There is a known issue related to clusters not having a default gateway making the nncp always to rollback, there are plans to use different approach like checking name servers.
+
+he desired state fields follow the nmstate API described at their [awesome doc](https://www.nmstate.io/)
```
```suggestion
The desired state fields follow the nmstate API described at their [awesome doc](https://www.nmstate.io/)
```

RamLavi commented on _posts/2020-09-23-attaching-to-multiple-networks-followup.markdown:
```diff
@@ -0,0 +1,618 @@
+
+Also for more details on kubernetes-nmstate there are guides covering [reporting](https://github.com/nmstate/kubernetes-nmstate/blob/master/docs/user-guide-101-reporting.md), [configuration](https://github.com/nmstate/kubernetes-nmstate/blob/master/docs/user-guide-102-configuration.md) and [troubleshooting](https://github.com/nmstate/kubernetes-nmstate/blob/master/docs/user-guide-103-troubleshoot ...
+
+## Demo: mixing it all together, VMI to VMI communication between nodes
```
```suggestion
## Demo: mixing it all together, VM to VM communication between nodes
```

can you also do a sweep and make sure `vmi` is not used when `vm` should be?

RamLavi commented on _posts/2020-09-23-attaching-to-multiple-networks-followup.markdown:
```diff
@@ -0,0 +1,618 @@
+apiVersion: kubevirt.io/v1alpha3
+kind: VirtualMachine
+metadata:
+  name: vma
```
do we want to set explicit namespace to avoid HCO setting it on HCO's default ns?

qinqon commented:
Push:
* Addressed Ram's comments

RamLavi commented on _posts/2020-09-23-attaching-to-multiple-networks-followup.markdown:
````diff
@@ -0,0 +1,618 @@
+
+```yaml
+cat <<EOF | kubectl apply -f -
+apiVersion: nmstate.io/v1alpha1
````
shouldn't it be [v1beta1](https://github.com/nmstate/kubernetes-nmstate/blob/master/deploy/crds/nmstate.io_nodenetworkconfigurationpolicies.yaml)?

qinqon replied:
Nope, latest released CNAO does not has that yet, https://github.com/kubevirt/cluster-network-addons-operator/blob/v0.42.2/data/nmstate/nmstate.io_nodenetworkconfigurationpolicies_crd.yaml.

RamLavi commented:
/lgtm

rhrazdil commented on _posts/2020-09-23-attaching-to-multiple-networks-followup.markdown:
```diff
@@ -0,0 +1,618 @@
+
+## Conclusion
+
+At this blog post we used some network components from kubevirt project to communicate two VMs at different nodes
```
In this blog post we used network components from kubevirt project to connect two VMs on different nodes

rhrazdil commented on _posts/2020-09-23-attaching-to-multiple-networks-followup.markdown:
```diff
@@ -0,0 +1,618 @@
+## Conclusion
+
+At this blog post we used some network components from kubevirt project to communicate two VMs at different nodes
+through a linux bridge connected to a secondary NIC, this illustrate how VM traffic can be connected to a specific NIC
```
through a linux bridge connected to a secondary NIC. This illustrates how VM traffic can be directed to a specific NIC

rhrazdil commented on _posts/2020-09-23-attaching-to-multiple-networks-followup.markdown:
```diff
@@ -0,0 +1,618 @@
+
+At this blog post we used some network components from kubevirt project to communicate two VMs at different nodes
+through a linux bridge connected to a secondary NIC, this illustrate how VM traffic can be connected to a specific NIC
+at node using a secondary NIC.
```
on a node using a secondary NIC on a VM.

phoracek commented on _posts/2020-09-23-attaching-to-multiple-networks-followup.markdown:
````diff
@@ -0,0 +1,618 @@
+    bridge.network.kubevirt.io/br0: 1k
+```
+
+This is an example of NetworkAttachmentDefinition to create a L2 bridge
````
It does not create a bridge, it exposes the bridge available on the host to users

phoracek commented on _posts/2020-09-23-attaching-to-multiple-networks-followup.markdown:
```diff
@@ -0,0 +1,618 @@
+
+To ensure the configuration is applied only in pods on nodes that have the bridge, the `k8s.v1.cni.cncf.io/resourceName` label is added. This goes hand in hand with another component, [bridge-marker](https://github.com/kubevirt/bridge-marker) which inspects nodes networking and if a new bridge pops up it will mark the node status with it.
+
+This is an example of the results from bridge-marker on nodes:
```
"on nodes where bridge br0 is already configured"

So it is clear that somebody needs to manage them. And in the next section you describe how.

phoracek commented on _posts/2020-09-23-attaching-to-multiple-networks-followup.markdown:
```diff
@@ -0,0 +1,618 @@
+      description: Linux bridge with eth1 as a port
+      type: linux-bridge
+      state: up
+      ipv4:
```
Can we drop this if it is not used? And if not, we should be able to drop at least the dhcp part since we set it to disabled anyway.

phoracek commented on _posts/2020-09-23-attaching-to-multiple-networks-followup.markdown:
```diff
@@ -0,0 +1,618 @@
+
+At this point we have ready L2 linux bridge connected to NIC eth1.
+
+### Configure multus attachment with a L2 bridge and a vlan
```
It is a network attachment. Multus just happens to be an implementation of this "de-facto standard"

phoracek commented on _posts/2020-09-23-attaching-to-multiple-networks-followup.markdown:
```diff
@@ -0,0 +1,618 @@
+br0. They will also have a secondary NIC eth1 to connect to the other VM running at different node, so they go
+over the br0 at nodes.
+
+The following picture illustrate the cluster:
```
illustrates

phoracek commented on _posts/2020-09-23-attaching-to-multiple-networks-followup.markdown:
````diff
@@ -0,0 +1,618 @@
+
+```bash
+PING 10.200.0.2 (10.200.0.2): 56 data bytes
+64 bytes from 10.200.0.2: seq=0 ttl=50 time=357.040 ms
````
Are we really getting this horrible latency?

qinqon replied:
Well this is two nodes kubevirtci at my laptop.

qinqon commented on _posts/2020-09-23-attaching-to-multiple-networks-followup.markdown:
```diff
@@ -0,0 +1,618 @@
+navbar_active: Blogs
+pub-date: October 13
+pub-year: 2020
+category: uncategorized
```
we have: `releases`, `updates`, `news` and `uncategorized`, they are pretty general categories maybe `updates` will fit ?

codificat replied:
For now please use `news`. While it doesn't really fit, it's what blog posts have been using...

We will improve this once issue #510 gets addressed

qinqon commented:
Adressed Radim's comments.

qinqon commented:
Adressed Petr's comments.

rhrazdil commented:
@qinqon Looks good to me

qinqon commented:
@codificat moved it to `news`

phoracek commented:
/lgtm

Thanks, awesome!

codificat commented on _posts/2020-09-23-attaching-to-multiple-networks-followup.markdown:
```diff
@@ -0,0 +1,615 @@
+title: Multiple Network Attachments with bridge CNI
+description: This post illustrates configuring secondary interfaces at VMs with a L2 linux bridge at nodes using just kube api.
+navbar_active: Blogs
+pub-date: October 13
```
Can you update this to `October 21` and rename the file to `2020-10-21-attaching...` accordingly please? Otherwise it appears as published on Sept 23rd

codificat commented on _posts/2020-09-23-attaching-to-multiple-networks-followup.markdown:
```diff
@@ -0,0 +1,615 @@
+
+## Introduction
+
+Over the last years KubeVirt project has improved a lot regarding secondary interfaces networking configuration. Now it's possible to do an end to end configuration from host networking to a VM using just the kubernetes API with
```
_the_ KubeVirt project

and _Kubernetes_ (capital K)

codificat commented on _posts/2020-09-23-attaching-to-multiple-networks-followup.markdown:
```diff
@@ -0,0 +1,615 @@
+## Introduction
+
+Over the last years KubeVirt project has improved a lot regarding secondary interfaces networking configuration. Now it's possible to do an end to end configuration from host networking to a VM using just the kubernetes API with
+special CRDs. Moreover, the deployment of all the projects has been simplified by introducing [kubevirt hyperconverged cluster operator (HCO)](https://github.com/kubevirt/hyperconverged-cluster-operator) and [cluster network addons operator (CNAO)](https://github.com/kubevirt/cluster-network-addons-operator) to install the networking components.
```
_KubeVirt_ (camel case) hyperconverged...

Also it's probably worth expanding the acronym _Custom Resource Definitions (CRDs)_ even if it's quite common these days.

codificat commented on _posts/2020-09-23-attaching-to-multiple-networks-followup.markdown:
```diff
@@ -0,0 +1,615 @@
+The [cluster network addons operator](https://github.com/kubevirt/cluster-network-addons-operator) manages the lifecycle (deploy/update/delete) of different k8s network components needed to
+configure secondary interfaces, manages MAC addresses and defines networking on hosts for pods and VMs.
+
+Good thing about having an operator is that everything is done through the API and you don't have to go over all nodes to install these components yourself and assures smooth updates.
```
_A_ good thing, or _The_ good thing it you prefer

codificat commented on _posts/2020-09-23-attaching-to-multiple-networks-followup.markdown:
```diff
@@ -0,0 +1,615 @@
+## Introducing cluster-network-addons-operator
+
+The [cluster network addons operator](https://github.com/kubevirt/cluster-network-addons-operator) manages the lifecycle (deploy/update/delete) of different k8s network components needed to
+configure secondary interfaces, manages MAC addresses and defines networking on hosts for pods and VMs.
```
_manage_ [...] _define_  without the "s", for sentence coherence and also because it's not CNAO who does that but the components it installs (right?)

codificat commented on _posts/2020-09-23-attaching-to-multiple-networks-followup.markdown:
```diff
@@ -0,0 +1,615 @@
+- kubemacpool: to manage mac addresses
+- kubernetes-nmstate: to configure the linux bridge on the nodes
+
+The list of components we want CNAO to deploy is specified by the NetworkAddonsConfig CR and the progress of the installation appears in the CR status field, split per component. To inspect
```
First use acronym expansion: _Custom Resource (CR)_

Also we normally backtick `NetworkAddonsConfig` for formatting of code-like references

codificat commented on _posts/2020-09-23-attaching-to-multiple-networks-followup.markdown:
```diff
@@ -0,0 +1,615 @@
+
+## Connecting Pods, VMs and Nodes over a single secondary network with bridge CNI
+
+Although kubernetes provides a default interface that gives connectivity to pods and VMs, it's not easy to configure which NIC should be used for specific pods or VMs in a multi NIC node cluster. Typical use case is to split control/traffic planes isolated by different NICs on nodes.
```
*K*ubernetes.

_A_ typical use case

codificat commented on _posts/2020-09-23-attaching-to-multiple-networks-followup.markdown:
````diff
@@ -0,0 +1,615 @@
+    bridge.network.kubevirt.io/br0: 1k
+```
+
+This is an example of NetworkAttachmentDefinition to exposes the bridge available on the host to users:
````
Backtick `NetworkAttachmentDefinition`

Also: _to expose_ (without "s")

codificat commented on _posts/2020-09-23-attaching-to-multiple-networks-followup.markdown:
````diff
@@ -0,0 +1,615 @@
+    }
+```
+
+Then adding bridge secondary network to a pod is a matter of adding the following annotation to
````
Then adding _the_ bridge...

codificat commented on _posts/2020-09-23-attaching-to-multiple-networks-followup.markdown:
```diff
@@ -0,0 +1,615 @@
+
+## Setting up node networking with NodeNetworkConfigurationPolicy (aka nncp)
+
+Changing kubernetes cluster node networking can be done manually iterating over all the cluster nodes and making changes or using different automatization tools like ansible. However, using just another kubernetes resource is more convenient.
```
*K*ubernetes (x3)

codificat commented on _posts/2020-09-23-attaching-to-multiple-networks-followup.markdown:
```diff
@@ -0,0 +1,615 @@
+Changing kubernetes cluster node networking can be done manually iterating over all the cluster nodes and making changes or using different automatization tools like ansible. However, using just another kubernetes resource is more convenient.
+For this purpose the kubernetes-nmstate project was born as a cluster wide node network administrator based on kubernetes CRs on top of [nmstate](https://github.com/nmstate/nmstate).
+
+It works as a k8s DaemonSet running pods on all the cluster nodes and reconciling three different CRs:
```
s/k8s/Kubernetes/

`DaemonSet`

codificat commented on _posts/2020-09-23-attaching-to-multiple-networks-followup.markdown:
```diff
@@ -0,0 +1,615 @@
+> note "Note"
+> Project kubernetes-nmstate has a distributed architecture to reduce kube-apiserver connectivity dependency, this means that every pod will configure the networking on the node that it's running without much interaction with kube-apiserver.
+
+In case something goes wrong and the pod changing the node network cannot ping the default gateway or has lost the kube-apiserver connectivity it will rollback to the previous configuration to go back to a working state. Those errors can be checked by running `kubectl get nnce`. The command displays potential issue per node and nncp.
```
potential issue*s* (missing s)

codificat commented on _posts/2020-09-23-attaching-to-multiple-networks-followup.markdown:
```diff
@@ -0,0 +1,615 @@
+In case something goes wrong and the pod changing the node network cannot ping the default gateway or has lost the kube-apiserver connectivity it will rollback to the previous configuration to go back to a working state. Those errors can be checked by running `kubectl get nnce`. The command displays potential issue per node and nncp.
+
+> note "Note"
+> There is a known issue related to clusters not having a default gateway making the nncp always to rollback, there are plans to use different approach like checking name servers.
```
Are there links available to the issue / plans?

qinqon replied:
We have just merge the fix https://github.com/nmstate/kubernetes-nmstate/pull/606, I will remove the note.

codificat commented on _posts/2020-09-23-attaching-to-multiple-networks-followup.markdown:
```diff
@@ -0,0 +1,615 @@
+
+## Demo: mixing it all together, VM to VM communication between nodes
+
+With the following recipe we will end up with a virtual machines pair on two different nodes with one secondary NICs, eth1 at vlan 100. They will be connected to each other using
```
... end up with a pair of virtual machines ...

codificat commented on _posts/2020-09-23-attaching-to-multiple-networks-followup.markdown:
```diff
@@ -0,0 +1,615 @@
+With the following recipe we will end up with a virtual machines pair on two different nodes with one secondary NICs, eth1 at vlan 100. They will be connected to each other using
+the same bridge on nodes that also have the external secondary NIC eth1 connected.
+
+### Requirements
```
Are these real requirements? I think they are not (e.g. kubevirtci is used for illustration purposes)... so I would remove that headline, or better change it to something like "demo environment setup" to make it more clear

codificat commented on _posts/2020-09-23-attaching-to-multiple-networks-followup.markdown:
````diff
@@ -0,0 +1,615 @@
+```
+
+We can inspect the status of applying the policy to each node.
+For that there is the NodeNetworkConfigurationEnactment CR (nnce):
````
Backtick `NodeNetworkConfigurationEnactment`

codificat commented on _posts/2020-09-23-attaching-to-multiple-networks-followup.markdown:
```diff
@@ -0,0 +1,615 @@
+> In case of errors it is possible to retrieve the error dumped by nmstate running
+> `kubectl get nnce -o yaml` the status will contain the error.
+
+We can also inspect the network state on the nodes by retrieving the NodeNetworkState and
```
Backtick `NodeNetworkState` and maybe add "(nns)" next to it

codificat commented on _posts/2020-09-23-attaching-to-multiple-networks-followup.markdown:
````diff
@@ -0,0 +1,615 @@
+    bridge.network.kubevirt.io/br0: 1k
+```
+
+At this point we have ready L2 linux bridge connected to NIC eth1.
````
... we have an L2 linux bridge ready and connected...

codificat commented on _posts/2020-09-23-attaching-to-multiple-networks-followup.markdown:
```diff
@@ -0,0 +1,615 @@
+
+### Configure network attachment with a L2 bridge and a vlan
+
+In order to make the bridge a L2 bridge with vlan, we specify no ipam and
```
Something is unclear in that sentence, at least to me... could you expand it a bit to help clarify?

Besides that, *IPAM* (capitals for acronym) also probably deserves expansion and/or claritication... why does (no) IPAM apply to the VLAN case and it doesn't it there is no VLAN?

qinqon replied:
Splitted at two stences one for L2 and the other for vlan, since IPAM and VLAN has no relation.

codificat commented on _posts/2020-09-23-attaching-to-multiple-networks-followup.markdown:
````diff
@@ -0,0 +1,615 @@
+sudo install virtctl /usr/local/bin
+```
+
+Now let's create two VirtualMachines on each node they will have one secondary NIC connected to br0 using the multus configuration for vlan 100. We will also activate kubemacpool to be sure that mac addresses are unique in the cluster and install the qemu-guest-agent so ip addresses from secondary NICs are reported to VM and we can inspect them later on.
````
Backtick `VirtualMachine`s.
Missing period (.) to separate sentences: ... on each node. They will...

Also *IP* (capitals)

codificat commented on _posts/2020-09-23-attaching-to-multiple-networks-followup.markdown:
```diff
@@ -0,0 +1,615 @@
+
+## Conclusion
+
+In this blog post we used network components from kubevirt project to connect two VMs on different nodes
```
... from *the KubeVirt* project...

codificat reviewed (changes requested):
Great post, thanks!

Only a few minor comments around syntax/grammar mostly.

qinqon commented:
push: Addressed @codificat comments

codificat commented:
Great, thanks!
/approve

codificat commented:
oh, and
/lgtm
